http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala

```diff
@@ -782,31 +787,56 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
       for {
         ctx <- ev.ask
         wsmApi <- buildWsmControlledResourceApiClient
-        controlledDbsForApp = appInstall.databases.collect { case d @ ControlledDatabase(_, _, _) => d }
-        // retrieve databases that might already be created in workspace
+
+        // get a list of databases needed for this app
+        controlledDbsRequiredForApp = appInstall.databases.collect { case d @ ControlledDatabase(_, _, _) => d }
+
+        // retrieve set of info about any WSM existing databases for this workspace
         existingControlledDbsInWorkspace <- retrieveWsmDatabases(wsmResourceApi,
-                                                                 controlledDbsForApp.map(_.prefix).toSet,
+                                                                 controlledDbsRequiredForApp.map(_.prefix).toSet,
                                                                  workspaceId.value
         )
-        wsmControlledDBResources <- controlledDbsForApp
+        wsmControlledDBResources <- controlledDbsRequiredForApp
           .map { controlledDbForApp =>
+            val maybeExistingDb = existingControlledDbsInWorkspace
+              .find(existingDb => controlledDbForApp.prefix == existingDb.wsmDatabaseName)
+
             // if a database already exists (because of workspace cloning) use that otherwise create a new one
-            if (
-              existingControlledDbsInWorkspace
-                .exists(existingDb => controlledDbForApp.prefix == existingDb.wsmDatabaseName)
-            ) {
-              F.pure(
-                existingControlledDbsInWorkspace
-                  .find(clonedDatabase => controlledDbForApp.prefix == clonedDatabase.wsmDatabaseName)
-                  .get
-              )
-            } else {
-              createWsmDatabaseResource(app, workspaceId, controlledDbForApp, namespacePrefix, owner, wsmApi).map {
-                db =>
-                  WsmControlledDatabaseResource(db.getAzureDatabase.getMetadata.getName,
-                                                db.getAzureDatabase.getAttributes.getDatabaseName
-                  )
-              }
+            maybeExistingDb match {
+              case Some(existingDb) =>
+                // there is an existing db, so check to see if we have previously mapped it to appControlledResource
+                val resourceId = existingDb.controlledResourceId.get
+                val maybeAppControlledResource = appControlledResourceQuery
+                  .getAllForAppByResourceId(app.id.id, WsmControlledResourceId(resourceId))
+                  .transaction
+
+                // if the appControlledResource already exists for this app, use
+                maybeAppControlledResource match {
+                  case Some(_) =>
+                    F.pure(existingDb)
+
+                  case Nil =>
```

Comment:
🤔 this is effectively doing the backfilling / migration. Instead of having this branch of logic. Would it make sense to do a one time backfilling manually, and then remove this case from the code altogether?

Reply:
The migration will only fix up existing cloned workspaces/databases that are missing the app_controlled_resource link. But when a new clone happens, this code links the APP_CONTROLLED_RESOURCE to the newly cloned database.  So I may be not understand or have it backward, but I thought that with this we may not have to do the migration since this will catch and fix the missing app_controlled_resource record. But, will this guaranteed to catch it before it is ever used/needed elsewhere?

Reply:
Okay, so the use cases here are:
1- There is an existing DB and it has been previously linked (cloned or not cloned)
2- There is an existing DB, but it has not been linked to the app (cloned)
3- There is no DB and we need to create one (not cloned)

If that's the case, adding these comments in the code would help a lot I think

Reply:
Unless I'm mistaken, this code path (`createOrFetchWsmDatabaseResources`) only gets called for _new_ app creations (from `createAndPollApp`). So this should fix the issue going forward, I think. But we still have existing apps who are missing the `APP_CONTROLLED_RESOURCE` records. We could either do a one-time migration script, or maybe add some code path from `updateAndPollApp` to fix up those records at upgrade tiem?